tests: Bluetooth: ascs: Add Source ASE state transition tests

MariuszSkamra opened the pull request:
This adds source ASE state transition tests and improves the callback call validation.

MariuszSkamra committed:
tests: Bluetooth: ascs: Set pacs_cap_foreach_custom_fake as default mock

This moves pacs_cap_foreach_custom_fake to mocks to be default mock.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
tests: Bluetooth: ascs: Reset mocks after test preamble

This adds test_mocks_reset function to be called after test case
preamble. The motivation is to reset the function call state that might
be affected by steps perfromed in the preamble stage, which is not the
code under test.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
tests: Bluetooth: ascs: Fix invalid NULL assignment

The iso_server shall be not set to NULL in mock_bt_iso_cleanup(), as
it's UUT responsibility to call bt_iso_server_unregister(). Cleaning
the iso_server in mock_bt_iso_cleanup() might hide bug in the code.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
tests: Bluetooth: ascs: Fix missing inline

This fixes missing inline function specifier.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
tests: Bluetooth: ascs: Fix minor typo

This fixes invalid value comparison argument.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
tests: Bluetooth: ascs: Add Source ASE state transition tests

This adds source ASE state transition tests.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: ascs: Fix missing ISO state check

This fixes missing ISO connection state check, as there might be no ISO
connection when Receiver Stop Ready has been received. This might happen
on ASE state transition from Enabling to Disabling.
Without this check the code asserts in ascs_disconnect_stream_work_handler.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -2376,10 +2376,12 @@ static void ase_stop(struct bt_ascs_ase *ase)
-	if (err < 0) {
-		LOG_ERR("Failed to disconnect stream %p: %d", stream, err);
-		return;
+	if (ep->iso != NULL && ep->iso->chan.state != BT_ISO_STATE_DISCONNECTED) {
```
May be worth to also check for the `BT_ISO_STATE_DISCONNECTING` state

Thalley commented on tests/bluetooth/audio/ascs/src/test_ase_state_transition.c:
```diff
@@ -25,7 +25,10 @@
 #include "test_common.h"
 
-struct test_sink_ase_state_transition_fixture {
+#define test_sink_ase_state_transition_fixture test_ase_state_transition_fixture
+#define test_source_ase_state_transition_fixture test_ase_state_transition_fixture
```
What is the purpose of these?

MariuszSkamra replied:
Test suite's fixture has to be named after test suite followed by `_fixture` suffix.
Example: `my_test_suite` `my_test_suite_fixture`.
In this case I have 2 test suits: `test_sink_ase_state_transition` and `test_source_ase_state_transition`, that share common `struct test_ase_state_transition_fixture`. Thus I added those defines to make it compile :D

Thalley replied:
> Test suite's fixture has to be named after test suite followed by _fixture suffix.

Is that something that the framework requires? 

But yeah I guess this makes sense if that is required, and it's more maintainable than defining two identical structs with different names :)

Thalley commented on tests/bluetooth/audio/ascs/src/test_ase_state_transition.c:
```diff
@@ -93,10 +114,29 @@ ZTEST_F(test_sink_ase_state_transition, test_client_codec_configured_to_qos_conf
+
+	/* Verification */
 	expect_bt_bap_unicast_server_cb_qos_called_once(stream, EMPTY);
+	expect_bt_bap_unicast_server_cb_config_not_called();
+	expect_bt_bap_unicast_server_cb_reconfig_not_called();
+	expect_bt_bap_unicast_server_cb_enable_not_called();
+	expect_bt_bap_unicast_server_cb_start_not_called();
+	expect_bt_bap_unicast_server_cb_metadata_not_called();
+	expect_bt_bap_unicast_server_cb_disable_not_called();
+	expect_bt_bap_unicast_server_cb_stop_not_called();
+	expect_bt_bap_unicast_server_cb_release_not_called();
```
Hmm, it seems overkill to define which functions are not called, instead of just the ones that are called. 

Could we not define these test so by default it expects none of them called, and then only define which should be called here?

MariuszSkamra replied:
Sounds reasonable

Thalley commented on tests/bluetooth/audio/ascs/src/test_ase_state_transition.c:
```diff
@@ -365,17 +608,692 @@ ZTEST_F(test_sink_ase_state_transition, test_client_streaming_to_qos_configured)
+	Z_TEST_SKIP_IFNDEF(CONFIG_BT_ASCS_ASE_SRC);
+
+	/* Preamble */
+	test_ase_control_client_config_codec(conn, ase_id, stream);
+	test_ase_control_client_config_qos(conn, ase_id);
+	test_ase_control_client_enable(conn, ase_id);
```
Would it make sense to define a helper function which calls these functions based on the wanted state? So that we could just do
```
test_ase_control_client_set_preamble_state(BT_BAP_EP_STATE_ENABLING);
```

Thalley replied:
Because basically what this test really wants to do is
```
setup_preamble;
test_ase_control_client_release(conn, ase_id);

expect_bt_bap_unicast_server_cb_release_called_once(stream);
expect_bt_bap_stream_ops_released_called_once(stream);
```

but instead of being 4 (or maybe a few more) lines, it's 34. It's not necessarily bad to be verbose in testing, but I'm concerned about the number of lines that are essentially copied for each new test case, which could cause developers to forget one or 2.

MariuszSkamra committed:
tests: Bluetooth: ascs: Rework expect_bt_bap_stream_ops_stopped_called_once

Make the parameters optional allowing to skip the function parameter
verification.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
tests: Bluetooth: ascs: Factor out test preambles

Factor out test preambles, to deduplicate the code and improve it's
readablility.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra commented:
rebased

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -2376,10 +2376,14 @@ static void ase_stop(struct bt_ascs_ase *ase)
-	if (err < 0) {
-		LOG_ERR("Failed to disconnect stream %p: %d", stream, err);
-		return;
+	if (ep->iso != NULL &&
+	    ep->iso->chan.state != BT_ISO_STATE_DISCONNECTED &&
+	    ep->iso->chan.state != BT_ISO_STATE_DISCONNECTING) {
+		err = ascs_disconnect_stream(stream);
+		if (err < 0) {
+			LOG_ERR("Failed to disconnect stream %p: %d", stream, err);
+			return;
+		}
```
For the unicast client we actually have a check like `unicast_client_can_disconnect_stream` which is similar to this, but also takes into account that a CIS may be used by 2 audio streams. I guess that last part might still be missing here, right? i.e. if the CIS is used for another stream in the other direction that is still streaming, we should not attempt to disconnect the CIS here, right?

MariuszSkamra commented:
Rebased
